# Tokenization :
##  * Tokenization is at the heart of much weirdness of LLMs. Do not brush it off.

   + Why can't LLM spell words? Tokenization.
   
   
   + Why can't LLM do super simple string processing tasks like reversing a string? Tokenization.
   
   
   + Why is LLM worse at non-English languages (e.g. Japanese)? Tokenization.
   
   
   + Why is LLM bad at simple arithmetic? Tokenization.
   
   
   + Why did GPT-2 have more than necessary trouble coding in Python? Tokenization.
   
   
   + Why did my LLM abruptly halt when it sees the string "<|endoftext|>"? Tokenization.
   
   
   + What is this weird warning I get about a "trailing whitespace"? Tokenization.
   
   
   + Why the LLM break if I ask it about "SolidGoldMagikarp"? Tokenization.
   
   
   + Why should I prefer to use YAML over JSON with LLMs? Tokenization.
   
   
   + Why is LLM not actually end-to-end language modeling? Tokenization.
   
   
   + What is the real root of suffering? Tokenization.


# -----------------------------


```python 

 Tokenization is at the heart of much weirdness of LLMs. Do not brush it off.
 
 127 + 677 = 804
 1275 + 6773 = 8041
 
 Egg.
 I have an Egg.
 egg.
 EGG.
 
 만나서 반가워요. 저는 OpenAI에서 개발한 대규모 언어 모델인 ChatGPT입니다. 궁금한 것이 있으시면 무엇이든 물어보세요.
 
 for i in range(1, 101):
     if i % 3 == 0 and i % 5 == 0:
         print("FizzBuzz")
     elif i % 3 == 0:
         print("Fizz")
     elif i % 5 == 0:
         print("Buzz")
     else:
         print(i)
```

# -----------------------------

#### * So thsis is sentence with 3 diff, Str,:
  * First its Korean, 
  * and then Specail Char, :Emoj 👋
  * finaly English

  + So now we need to feed this to TransF.

In [ ]:
"안녕하세요 👋 (hello in Korean!)" 

'안녕하세요 👋 (hello in Korean!)'

# What is Str. in python Anyway..?

  + in Py. Doc.;
    * Textual data in Python is handled with str objects, or strings.,
    
    *  **Strings are immutable sequences of Unicode code points**. String literals are written in a variety of ways:


# Unicodes :
  
   * Unicode, formally The **Unicode Standard**, 
   
   * is a **text encoding standard** maintained by the Unicode Consortium designed to support the use of text in all of the **world's writing systems that can be digitized**.


#### In python we can access, unicode of a str, Using **ord**  



In [ ]:
# Access, all the unicode for Chars in sentence
[ord(x) for x in '안녕하세요 👋 (hello in Korean!)'] 

[50504,
 45397,
 54616,
 49464,
 50836,
 32,
 128075,
 32,
 40,
 104,
 101,
 108,
 108,
 111,
 32,
 105,
 110,
 32,
 75,
 111,
 114,
 101,
 97,
 110,
 33,
 41]

In [5]:
ord('A')

65

# UTF-8 ,/
  
  + UTF-8 is a **character encoding standard used for electronic communication**. 
  
  + Defined by the Unicode Standard, the name is derived from **Unicode Transformation Format – 8-bit**. 
  
  + Almost every webpage is stored in UTF-8.


### We Tend to use ``utf-8`` enc, Cos. Utf-16 and Utf-32, quite Wasting,

### They have lot of 0 in enc, so in our case utf-8 is ok.

In [2]:
# Conver to utf-8
list('안녕하세요 👋 (hello in Korean!)'.encode('utf-8'))

[236,
 149,
 136,
 235,
 133,
 149,
 237,
 149,
 152,
 236,
 132,
 184,
 236,
 154,
 148,
 32,
 240,
 159,
 145,
 139,
 32,
 40,
 104,
 101,
 108,
 108,
 111,
 32,
 105,
 110,
 32,
 75,
 111,
 114,
 101,
 97,
 110,
 33,
 41]

### But we cant just Feed this row Byte, input to a TransF..:

  * So we have to Compres it using Byte Pair Encoding Algorithm

# -----

# Byte Pair Encoding
  
  + Byte pair encoding, (also known as digram coding) is an algorithm, first described in **1994** by **Philip Gage** 
  
  + for **encoding strings of text into tabular** form for **use in downstream modeling**.

  + Its modification is notable as the large LM tokenizer, with an ability to **combine both tokens that encode single characters (including single digits or single punctuation marks)** 
  
  + and those that encode whole words (even the longest compound words).


## Original algorithm.
  
  + The original algorithm operates by **iteratively replacing the most common contiguous sequences of characters** in a **target text** with unused ``'placeholder'`` bytes.

   * Example:
      + Suppose the data to be encoded is
             
                    aaabdaaabac
      
      + The byte pair ``"aa"`` occurs most often, **so it will be replaced by a byte that is not used in the data, such as "Z".** Now there is the following data and replacement table:

                    ZabdZabac
                    Z =aa


### * So, we got, seq of only ``9`` token, but now with the voab of ``5``, Cos we got 5th vocab element ``Z``. to cat aa

### * And we can continue Doing this,, by cat common chars. 

### * By cat, we ``decreas the token``, ``Increase the Vocab Siz``

In [4]:
txt = "Ｕｎｉｃｏｄｅ! 🅤🅝🅘🅒🅞🅓🅔‽ 🇺‌🇳‌🇮‌🇨‌🇴‌🇩‌🇪! 😄 The very name strikes fear and awe into the hearts of programmers worldwide. We all know we ought to “support Unicode” in our software (whatever that means—like using wchar_t for all the strings, right?). But Unicode can be abstruse, and diving into the thousand-page Unicode Standard plus its dozens of supplementary annexes, reports, and notes can be more than a little intimidating. I don’t blame programmers for still finding the whole thing mysterious, even 30 years after Unicode’s inception."
token = txt.encode('utf-8')
token = list(map(int, token))

print('=======')
print(txt)
print('Len of Txt',len(txt))


print('=======')
print(token)
print('Len of Tken',len(token))


Ｕｎｉｃｏｄｅ! 🅤🅝🅘🅒🅞🅓🅔‽ 🇺‌🇳‌🇮‌🇨‌🇴‌🇩‌🇪! 😄 The very name strikes fear and awe into the hearts of programmers worldwide. We all know we ought to “support Unicode” in our software (whatever that means—like using wchar_t for all the strings, right?). But Unicode can be abstruse, and diving into the thousand-page Unicode Standard plus its dozens of supplementary annexes, reports, and notes can be more than a little intimidating. I don’t blame programmers for still finding the whole thing mysterious, even 30 years after Unicode’s inception.
Len of Txt 533
[239, 188, 181, 239, 189, 142, 239, 189, 137, 239, 189, 131, 239, 189, 143, 239, 189, 132, 239, 189, 133, 33, 32, 240, 159, 133, 164, 240, 159, 133, 157, 240, 159, 133, 152, 240, 159, 133, 146, 240, 159, 133, 158, 240, 159, 133, 147, 240, 159, 133, 148, 226, 128, 189, 32, 240, 159, 135, 186, 226, 128, 140, 240, 159, 135, 179, 226, 128, 140, 240, 159, 135, 174, 226, 128, 140, 240, 159, 135, 168, 226, 128, 140, 240, 159, 135, 180, 226, 128, 140, 240

# Lest apply Byte Pair Encoding Algo. for this txt

# ===============================
##### I implement some func to replace same chars but this isnt what we want.. 

In [ ]:
"""
counts = {}
for i in range(100):
    for pair in zip(txt, txt[1:]):
        counts[pair] = counts.get(pair, 0) + 1
        if counts is None:
            mxCountPair = max(counts, key=counts.get)
"""

"""
txt = 'aaab ss s a a  dls s '
couts = {}
for i in range(100):
    for pair in zip(txt, txt[1:]):
        counts[pair] = counts.get(pair, 0) + 1
    if not counts:
        break
    maxCountPair = max(counts, key=counts.get)
    # Wrong: counts[maxCountPair] = 'ZZ'
    symbol = f'X{i}'
    txt = txt.replace(''.join(maxCountPair), symbol)
    counts = {}
"""
#print(txt)
#print(counts)

TypeError: '>' not supported between instances of 'int' and 'str'

In [10]:
import random

for i in range(10):
    counts = {}
    for pair in zip(txt, txt[1:]):
        counts[pair] = counts.get(pair, 0) + 1
    
    if not counts:
        break

    maxCountPair = max(counts, key=counts.get) 

    a = ['E', 'W', 'X', 'A']
    symbol = f'{random.choice(a)}{i}'

    txt = txt.replace(''.join(maxCountPair), symbol)

print("Final Text:", txt)
print("Final Counts:", counts)



Final Text: Ｕｎｉｃｏｄｅ! 🅤🅝🅘🅒🅞🅓🅔‽ 🇺‌🇳‌🇮‌🇨‌🇴‌🇩‌🇪! 😄 ThE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2very namE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2strikes fear and awE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2into thE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2hearts of programmers worldwide. WE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2all know wE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2ought to “support Unicode” in our softwarE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2(whatever that means—likE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2using wchar_t for all thE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2strings, right?). But UnicodE9ADE91E97A6ADE91E97A6E9329A1A197E95ADE91E97A6E932ADE91E97A6ADE91E97A6E93X2can bE

# ===============================

In [5]:
def getStats(ids):
    counts = {}
    for pair in zip(ids, ids[1:]):
        counts[pair] = counts.get(pair, 0) + 1
    return counts

stats = getStats(token)
print(sorted(((v, k) for k, v in stats.items()), reverse=True))

[(20, (101, 32)), (15, (240, 159)), (12, (226, 128)), (12, (105, 110)), (10, (115, 32)), (10, (97, 110)), (10, (32, 97)), (9, (32, 116)), (8, (116, 104)), (7, (159, 135)), (7, (159, 133)), (7, (97, 114)), (6, (239, 189)), (6, (140, 240)), (6, (128, 140)), (6, (116, 32)), (6, (114, 32)), (6, (111, 114)), (6, (110, 103)), (6, (110, 100)), (6, (109, 101)), (6, (104, 101)), (6, (101, 114)), (6, (32, 105)), (5, (117, 115)), (5, (115, 116)), (5, (110, 32)), (5, (100, 101)), (5, (44, 32)), (5, (32, 115)), (4, (116, 105)), (4, (116, 101)), (4, (115, 44)), (4, (114, 105)), (4, (111, 117)), (4, (111, 100)), (4, (110, 116)), (4, (110, 105)), (4, (105, 99)), (4, (104, 97)), (4, (103, 32)), (4, (101, 97)), (4, (100, 32)), (4, (99, 111)), (4, (97, 109)), (4, (85, 110)), (4, (32, 119)), (4, (32, 111)), (4, (32, 102)), (4, (32, 85)), (3, (118, 101)), (3, (116, 115)), (3, (116, 114)), (3, (116, 111)), (3, (114, 116)), (3, (114, 115)), (3, (114, 101)), (3, (111, 102)), (3, (111, 32)), (3, (108, 108)), (

# ord() give you unicode Val -- > ord(a)-> 92 
# chr() give you char of unicode val -- > chr(102)->'a'

In [ ]:
chr(101), chr(32) # So e and space is the most Occuring chars

('e', ' ')

In [21]:
maxOcuringPair = max(stats, key=stats.get)
maxOcuringPair

(101, 32)

In [6]:
def merge(ids, pair, ix):
    newId = []
    i = 0
    while i < len(ids):
        # if we are not at the very last Position and The pair matches, Replace it 
        if i < len(ids) - 1 and ids[i] == pair[0] and ids[i+1] == pair[1]:
            newId.append(ix)
            i += 2
        else:
            newId.append(ids[i])
            i += 1
    return newId

print(merge([5, 6, 6, 7, 9, 1], (6, 6), 99))

    


[5, 99, 7, 9, 1]


In [29]:
mergToken = merge(token, maxOcuringPair, 256)
print(mergToken)
print(len(mergToken))

[239, 188, 181, 239, 189, 142, 239, 189, 137, 239, 189, 131, 239, 189, 143, 239, 189, 132, 239, 189, 133, 33, 32, 240, 159, 133, 164, 240, 159, 133, 157, 240, 159, 133, 152, 240, 159, 133, 146, 240, 159, 133, 158, 240, 159, 133, 147, 240, 159, 133, 148, 226, 128, 189, 32, 240, 159, 135, 186, 226, 128, 140, 240, 159, 135, 179, 226, 128, 140, 240, 159, 135, 174, 226, 128, 140, 240, 159, 135, 168, 226, 128, 140, 240, 159, 135, 180, 226, 128, 140, 240, 159, 135, 169, 226, 128, 140, 240, 159, 135, 170, 33, 32, 240, 159, 152, 132, 32, 84, 104, 256, 118, 101, 114, 121, 32, 110, 97, 109, 256, 115, 116, 114, 105, 107, 101, 115, 32, 102, 101, 97, 114, 32, 97, 110, 100, 32, 97, 119, 256, 105, 110, 116, 111, 32, 116, 104, 256, 104, 101, 97, 114, 116, 115, 32, 111, 102, 32, 112, 114, 111, 103, 114, 97, 109, 109, 101, 114, 115, 32, 119, 111, 114, 108, 100, 119, 105, 100, 101, 46, 32, 87, 256, 97, 108, 108, 32, 107, 110, 111, 119, 32, 119, 256, 111, 117, 103, 104, 116, 32, 116, 111, 32, 226, 128, 156

In [7]:
vocabSiz = 276
numMergs = vocabSiz - 256
ids = list(token) # save a copy of original list

mergs = {}
for i in range(numMergs):
    stats = getStats(ids)
    pair = max(stats, key=stats.get)
    ix = 256 + i
    print(f'mergin para: {pair}, to --> {ix}')
    ids = merge(ids, pair, ix)
    mergs[pair] = ix 


mergin para: (101, 32), to --> 256
mergin para: (240, 159), to --> 257
mergin para: (226, 128), to --> 258
mergin para: (105, 110), to --> 259
mergin para: (115, 32), to --> 260
mergin para: (97, 110), to --> 261
mergin para: (116, 104), to --> 262
mergin para: (257, 133), to --> 263
mergin para: (257, 135), to --> 264
mergin para: (97, 114), to --> 265
mergin para: (239, 189), to --> 266
mergin para: (258, 140), to --> 267
mergin para: (267, 264), to --> 268
mergin para: (101, 114), to --> 269
mergin para: (111, 114), to --> 270
mergin para: (116, 32), to --> 271
mergin para: (259, 103), to --> 272
mergin para: (115, 116), to --> 273
mergin para: (261, 100), to --> 274
mergin para: (32, 262), to --> 275


In [32]:
print('tokenLen ', len(token))
print('ids len ', len(ids))
print(f'compression raitio {len(token) / len(ids):.2f}X',)

tokenLen  616
ids len  451
compression raitio 1.37X


Note, the Tokenizer is a completely separate, independent module from the LLM. It has its own training dataset of text (which could be different from that of the LLM), on which you train the vocabulary using the Byte Pair Encoding (BPE) algorithm. It then translates back and forth between raw text and sequences of tokens. The LLM later only ever sees the tokens and never directly deals with any text.

# Decoding 
 
  + given a seq of intgers in the range of [0, vocabSiz], Whats next?>>

In [17]:
# token id to byte Object For that token
vocab = {ix:bytes([ix]) for ix in range(256)} 
for (p0, p1), ix in mergs.items():
    vocab[ix] = vocab[p0] + vocab[p1] 


def decod(ids):
    # given ids (list of int), return python str /
    tokens = b"".join(vocab[ix] for ix in ids)
    txt = tokens.decode('utf-8', errors='replace')   
    return txt 

## So if we didnt use this [[error=replace]], we could get unicodeDecodeError,
 
  + Cos This 128 isn't a standard Unicode Repr.

  + **Read More On Docs, About Eror Handling** 

In [18]:
print(decod([128]))

�


# Encoding 
 
  + given a str, return list of int ( the tokens )

  

In [28]:
def enc(txt):
    tokn = list(txt.encode('utf-8'))
    while len(tokn) >= 2:
        stats = getStats(tokn)
        ## if not pair that didnt emegi, assing it to inf
        pair = min(stats, key=lambda p: mergs.get(p, float('inf')))
        if pair not in mergs:
            break
        ix = mergs[pair]
        tokn = merge(tokn, pair, ix)
    return tokn      
print(enc('hello world'))

[104, 101, 108, 108, 111, 32, 119, 270, 108, 100]


In [29]:
print(decod(enc('hello words')))

hello words


In [30]:
txt2 = decod(enc(txt))
print(txt2 == txt)

True


In [31]:
valTxt = 'Your code has the right idea of using a mapping (itos) to decode IDs into characters, but there are some issues in your implementation. Here’s the explanation and fixes:'
valTxt2 = decod(enc(valTxt))
print(valTxt == valTxt2)

True


# Forced splits Using regex Patterns (GPT series)

### NOTE:
  
  + **\p{L} or \p{Letter}**: any kind of letter from any language

  
  + **\p{N} or \p{Number}**: any kind of numeric character in any script.

  
  + Patterns Explained:
      + s matches the character s.
        

      + ``'t `` matches the contraction ``n't (as in "can't" or "don't").``
        

        
      + ``'re``  matches the contraction ``'re (as in "you're").``
        

        
      + ``'ve``  matches the contraction ``'ve (as in "we've").``
        

        
      + ``'m `` matches the contraction ``'m (as in "I'm").``
        


      + ``'ll``  matches the contraction ``'ll (as in "you'll").``
        
        

      + ``'d `` matches the contraction ``'d (as in "he'd" or "she'd").``



      + ``?[^\s\p{L}\p{N}]+`` matches the constraction, ``Not Letters of Num. [Punctuation]``


      + ``\s+(?!\S)|`` matches the constraction, ``Whitespace, '    hi', Up to last so--> '  ', ' hi'``   



      + ``\s+`` mathes the constraciton , ``catch Whitespce``


In [6]:
import regex as re 

gpt2Pat = re.compile(r"""'s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+""")

In [15]:
print(re.findall(gpt2Pat,  "hello've"))# |'ve

print(re.findall(gpt2Pat, 'hellow worlds')) # \p{L}

print(re.findall(gpt2Pat, 'helo123 worlds')) #\p{N}

print(re.findall(gpt2Pat, 'how`s life')) # hard code for this ['], 

print(re.findall(gpt2Pat, 'helo.???!!')) # ?[^\s\p{L}\p{N}]+ ,,, Not letters or num.,, Catch Punctuation

print(re.findall(gpt2Pat, '            helo')) # \s+(?!\S)|\s+ ,, Whistapce


['hello', "'ve"]
['hellow', ' worlds']
['helo', '123', ' worlds']
['how', '`', 's', ' life']
['helo', '.???!!']
['           ', ' helo']


In [16]:
# Example
example = """
for i in range(1, 101):
    if i % 3 == 0 or i % 5 == 0:
        print("FizzBuzz")
    elif i % 3 == 0:
        print('Fizz')
    elif i % 5 == 0:
        print('Buzz')
    else:
        print(i)
"""

print(re.findall(gpt2Pat, example))

['\n', 'for', ' i', ' in', ' range', '(', '1', ',', ' 101', '):', '\n   ', ' if', ' i', ' %', ' 3', ' ==', ' 0', ' or', ' i', ' %', ' 5', ' ==', ' 0', ':', '\n       ', ' print', '("', 'FizzBuzz', '")', '\n   ', ' elif', ' i', ' %', ' 3', ' ==', ' 0', ':', '\n       ', ' print', "('", 'Fizz', "')", '\n   ', ' elif', ' i', ' %', ' 5', ' ==', ' 0', ':', '\n       ', ' print', "('", 'Buzz', "')", '\n   ', ' else', ':', '\n       ', ' print', '(', 'i', ')', '\n']


# Tiktoken

In [ ]:
import tiktoken

# GPT-2, Dose not Merge ``Spaces``
enc = tiktoken.get_encoding('gpt2')
print(enc.encode("   hello world!!!"))


# GPT-4, Merge Spaces
enc = tiktoken.get_encoding("cl100k_base")
print(enc.encode("   hello world!!!"))

# get Visual Understand in Tiktokenizwe website

[220, 220, 23748, 995, 10185]
[256, 24748, 1917, 12340]


#### Reference the GPT-2 Enc, Download the vocab.bpe and encoder.json, 

In [1]:
!wget https://openaipublic.blob.core.windows.net/gpt-2/models/1558M/vocab.bpe
!wget https://openaipublic.blob.core.windows.net/gpt-2/models/1558M/encoder.json 

--2024-11-24 15:50:20--  https://openaipublic.blob.core.windows.net/gpt-2/models/1558M/vocab.bpe
Resolving openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)... 57.150.97.129
Connecting to openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)|57.150.97.129|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 456318 (446K) [application/octet-stream]
Saving to: ‘vocab.bpe’

vocab.bpe           100%[===================>] 445.62K   344KB/s    in 1.3s    

2024-11-24 15:50:23 (344 KB/s) - ‘vocab.bpe’ saved [456318/456318]

--2024-11-24 15:50:23--  https://openaipublic.blob.core.windows.net/gpt-2/models/1558M/encoder.json
Resolving openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)... 57.150.97.129
Connecting to openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)|57.150.97.129|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1042301 (1018K) [application/json]
Saving t

In [2]:
import os, json 

with open('encoder.json', 'r') as file:
    enc = json.load(file)

with open('vocab.bpe', 'r', encoding='utf-8') as file:
    bpeData = file.read()

bpeMerges = [tuple(mergStr.split()) for mergStr in bpeData.split('\n')[1:-1] ]


# Special Tokens

In [ ]:
len(enc) # 256 raw byte tokens + 50,000 merges +1 special Token 50,257

50257

In [ ]:
# what is 57 token
enc['<|endoftext|>'] # 57 token

50256